### PR TITLE
Add payment processing using ECE on the Blocks checkout and cart pages

### LINKED ISSUE #8773: Integrate ECE for Payment Processing in the Blocks Checkout Page
Implement the Stripe Express Checkout Element (ECE) to process payments on the WooCommerce Blocks checkout page. This task involves integrating ECE in a way that it handles payment transactions during the checkout process.

### Expected Behavior:

- ECE successfully processes payments on the WooCommerce Blocks checkout page without disrupting the existing checkout flow.
- It's not expected for the ECE checkout to function in any other flows or within the shortcode checkout.

### Tasks:

- [ ] Integrate ECE with WooCommerce Blocks
- [ ] Implement Payment Processing Logic

### LINKED ISSUE #8867: Integrate ECE for Payment Processing in the Blocks Cart Page
Implement the Stripe Express Checkout Element (ECE) to process payments on the WooCommerce Blocks cart page. 

### Acceptance criteria:

- The ECE implementation must successfully process payments on the WooCommerce Blocks cart page without disrupting the existing checkout flow.
- The ECE implementation is not expected to function in any other flows or within the shortcode checkout.
- Ensure there are no regressions in Payment Request Buttons (PRB) when ECE is disabled.
- There are no regressions when no express checkout payment methods are active.

### PR DESCRIPTION
**Pull request by rafaelzaleski**
Closes #8773 
Closes #8867 

#### Changes proposed in this Pull Request

Add payment processing using Express Checkout Elements (ECE) on the Blocks checkout and cart pages.

#### Testing instructions

**Test that there are no regressions in Payment Request Buttons (PRB) when ECE is disabled:**

Pre-requisites:
- The feature flag is disabled (`npm run wp option update _wcpay_feature_stripe_ece 0`).
- Confirm that Apple Pay and Google Pay are enabled in the merchant settings.

Test:

1. [Checkout with Payment Request](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows#checkout-with-payment-request)

**Test the ECE checkout on the blocks cart page using a simple card:**

Pre-requisites:
- The feature flag is enabled (`npm run wp option update _wcpay_feature_stripe_ece 1`).
- Confirm that Apple Pay and Google Pay are enabled in the merchant settings.

Test:

1. Add some products to the cart.
2. Navigate to the blocks cart page and wait for the ECE button to load.
3. Complete the payment using the ECE button via Google Pay. Use a real card; however, your card will not be charged if the merchant is in test mode.
4. Ensure the payment is processed successfully and that you are redirected to the thank you page.
5. Review the order in the merchant backend and ensure all the data is correct.
6. Check the order details in the Stripe dashboard to ensure all information is correct.

**Test the ECE checkout in the blocks checkout page using a simple card:**

Pre-requisites:
- The feature flag is enabled (`npm run wp option update _wcpay_feature_stripe_ece 1`).
- Confirm that Apple Pay and Google Pay are enabled in the merchant settings.

Test:

1. Add some products to the cart.
2. Navigate to the blocks checkout page and wait for the ECE button to load.
3. Complete the payment using the ECE button via Google Pay. Use a real card; however, your card will not be charged if the merchant is in test mode.
4. Ensure the payment is processed successfully and that you are redirected to the thank you page.
5. Review the order in the merchant backend and ensure all the data is correct.
6. Check the order details in the Stripe dashboard to ensure all information is correct.

**Ensure that there are no regressions when no express checkout payment methods are active.**

Pre-requisites:
- The feature flag is enabled (`npm run wp option update _wcpay_feature_stripe_ece 1`).
- Disable all express checkout payment methods in the merchant settings.

Test:

1. Add some products to the cart.
2. Navigate to the blocks checkout page, blocks cart page, shortcode checkout page, shortcode cart page, and product pages. Confirm that no express checkout buttons are visible.
3. Complete a checkout using a card on the classic checkout. Verify that the checkout process is completed without any errors.
4. Repeat step 1 by adding some products to the cart.
5. Complete a checkout using a card on the blocks checkout. Ensure the checkout process is completed without any errors.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8884` or branch name `add/8773-ece-support-blocks-checkout-page` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=add/8773-ece-support-blocks-checkout-page) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 30a68144e93be39935d9bc749ad7c1dc5e0dec83
- Build time: 2024-05-31 14:07:29 UTC

_Note: the build is updated when a new commit is pushed to this PR._

**Comment by asumaran** [comment]
Wanted to mention the ECE buttons are not displayed if the available space is not wide enough to render the button. I had to us the "Full width" template to be able to see the ECE buttons. Switching to another theme other than `Storefront` also works.

Also, the ECE is not displayed if WooPay is disabled. Is this the intended behavior @bborman22?

**Comment by rafaelzaleski** [comment]
> Also, the ECE is not displayed if WooPay is disabled. Is this the intended behavior @bborman22?

Fixed in https://github.com/Automattic/woocommerce-payments/pull/8884/commits/30a68144e93be39935d9bc749ad7c1dc5e0dec83

**Comment by rafaelzaleski** [comment]
> Wanted to mention the ECE buttons are not displayed if the available space is not wide enough to render the button. I had to us the "Full width" template to be able to see the ECE buttons. Switching to another theme other than Storefront also works.

Since we're behind a feature flag, this can be addressed in a new issue.

File: client/express-checkout/blocks/hooks/use-express-checkout.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	setExpressPaymentError,
+} ) => {
+	const stripe = useStripe();
+	const elements = useElements();
+
+	const buttonOptions = {
+		paymentMethods: {
+			applePay: 'always',
+			googlePay: 'always',
+			link: 'auto',
```
Comment (asumaran) [inline client/express-checkout/blocks/hooks/use-express-checkout.js:26]: When testing never got Link as payment option.

File: client/express-checkout/utils/normalize.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+export const normalizeOrderData = ( event, paymentMethodId ) => {
+	const name = event?.billingDetails?.name;
+	const email = event?.billingDetails?.email ?? '';
+	const phone = event?.billingDetails?.phone ?? '';
+	const billing = event?.billingDetails?.address ?? {};
+	const shipping = event?.shippingAddress ?? {};
+	const fraudPreventionTokenValue = window.wcpayFraudPreventionToken ?? '';
+
+	return {
+		billing_first_name:
```
Comment (asumaran) [inline client/express-checkout/utils/normalize.js:39]: seems like a good opportunity for start using [Zod](https://zod.dev/) here. cc @reykjalin

**Review by asumaran — APPROVED** [review]
Tests well in Chrome and Safari.

Just a few notes:

- ✅ Apple Pay in Safari works.
- ✅ Google Pay in Safari works.
- ✅ Google Pay in Chrome works.
- ⚠️ It doesn't work without WooPay enabled.
- ⚠️ Only USA billing and shipping addressed worked for me. Peruvian address fail. It would be great to show an error to the user when that happens.
- ⚠️ Never got to see Link as payment option.

File: client/express-checkout/utils/normalize.js (hunk 2 above)
Comment (rafaelzaleski) [reply client/express-checkout/utils/normalize.js:39]: We'll convert the tokenized payment requests to ECE, so the final files will probably be [these](https://github.com/Automattic/woocommerce-payments/tree/develop/client/tokenized-payment-request/transformers).

File: client/express-checkout/blocks/hooks/use-express-checkout.js (hunk 1 above)
Comment (rafaelzaleski) [reply client/express-checkout/blocks/hooks/use-express-checkout.js:26]: Me neither. I'll include the documentation on how to display Link in the scope of [this spike](https://github.com/Automattic/woocommerce-payments/issues/8874). Thank you for the heads up.

**Pull request merged**